bigquery/gcloud/aio/bigquery/bigquery.py

```diff
@@ -503,3 +527,64 @@ async def __aenter__(self) -> 'Table':
 
     async def __aexit__(self, *args: Any) -> None:
         await self.close()
+
+
+class Dataset(BigqueryBase):
+    def __init__(self, dataset_name: Optional[str] = None,
```

Comment:
It'd be great to make this `dataset_name` field required to avoid needing to have runtime validation in the bulk of it's methods; anywhere we can have schema failures occur in static analysis makes the library that much easier to debug.

What do you think about moving `list_datasets()` to `BigqueryBase` to accomplish this? Since `Dataset` inherits from it anyway, that means users would be able to do `Dataset().list_datasets()` if they need, but could also use the base client itself in cases where they don't have even a single named dataset to reference upfront.

Another option could be to expose this as a method: eg. something like `list_datasets(client: BigqueryBase, ...)`. I think I prefer having it be a member function, but wanted to present this as a valid option as well. Thoughts?

Reply:
One downside of this approach is that `list_datasets()` method will also be available in `Table` and `Job` class, which works but does not really make sense. 

I am not that against the idea of making `dataset_name` option. After all in class `Job` we have this pattern as well, a `Job` object does not need to work with a specific Job, similarly a `Dataset`object does not necessarily need to operate on a specific dataset either.

Reply:
Ah, good point -- I completely missed that, since it looks like we don't have explicit validation in the `Job` class as you've done here. Cool, looks good as is!